src/index.js

```diff
@@ -22,40 +59,62 @@ function Walker (dir, options) {
   if (options.depthLimit > -1) this.rootDepth = this.root.split(path.sep).length + 1
   this.fs = options.fs || require('graceful-fs')
 }
-util.inherits(Walker, Readable)
 
-Walker.prototype._read = function () {
-  if (this.paths.length === 0) return this.push(null)
-  var self = this
-  var pathItem = this.paths[this.options.queueMethod]()
+util.inherits(Walker, Readable)
 
-  var statFunction = this.options.preserveSymlinks ? self.fs.lstat : self.fs.stat
+Walker.prototype._readPath = function () {
+  var _ref = _asyncToGenerator(function* (pathItem) {
 
-  statFunction(pathItem, function (err, stats) {
-    var item = { path: pathItem, stats: stats }
-    if (err) return self.emit('error', err, item)
+    var statFunction = this.options.preserveSymlinks ? this.fs.lstat : this.fs.stat
+    var self = this
 
-    if (!stats.isDirectory() || (self.rootDepth &&
-      pathItem.split(path.sep).length - self.rootDepth >= self.options.depthLimit)) {
-      return self.push(item)
-    }
+    return statFunction(pathItem, function (err, stats) {
+      var item = { path: pathItem, stats: stats }
 
-    self.fs.readdir(pathItem, function (err, pathItems) {
       if (err) {
-        self.push(item)
-        return self.emit('error', err, item)
+        self.emit('error', err, item)
+        return
       }
 
-      pathItems = pathItems.map(function (part) { return path.join(pathItem, part) })
-      if (self.options.filter) pathItems = pathItems.filter(self.options.filter)
-      if (self.options.pathSorter) pathItems.sort(self.options.pathSorter)
-      // faster way to do do incremental batch array pushes
-      self.paths.push.apply(self.paths, pathItems)
+      if (!stats.isDirectory() || self.rootDepth &&
+	  pathItem.split(path.sep).length - self.rootDepth >= self.options.depthLimit) {
+        return self.push(item)
+      }
 
-      self.push(item)
+      self.fs.readdir(pathItem, function (err, pathItems) {
+        if (err) {
+          self.emit('error', err, item)
+          return self.push(item)
```

Comment:
The order of `emit` and `push` has been changed so that we can return the result of the call to `push` directly